testcases.py

```diff
@@ -628,12 +642,84 @@ def get_paths(self):
         self._files = [self._generate_random_file(10 * MB)]
         return self._files
 
-    def check(self):
+    def check(self) -> TestResult:
         num_handshakes = self._count_handshakes()
         if num_handshakes != 1:
             logging.info("Expected exactly 1 handshake. Got: %d", num_handshakes)
-            return False
-        return self._check_version_and_files()
+            return TestResult.FAILED
+        if not self._check_version_and_files():
+            return TestResult.FAILED
+        return TestResult.SUCCEEDED
+
+
+class TestCaseKeyUpdate(TestCaseHandshake):
+    @staticmethod
+    def name():
+        return "keyupdate"
+
+    @staticmethod
+    def testname(p: Perspective):
+        if p is Perspective.CLIENT:
+            return "keyupdate"
+        return "transfer"
+
+    @staticmethod
+    def abbreviation():
+        return "U"
+
+    def get_paths(self):
+        self._files = [self._generate_random_file(3 * MB)]
+        return self._files
+
+    def check(self) -> TestResult:
+        if not self._keylog_file():
+            logging.info("Can't check test result. SSLKEYLOG required.")
+            return TestResult.UNSUPPORTED
+
+        num_handshakes = self._count_handshakes()
+        if num_handshakes != 1:
+            logging.info("Expected exactly 1 handshake. Got: %d", num_handshakes)
+            return TestResult.FAILED
+        if not self._check_version_and_files():
+            return TestResult.FAILED
+
+        client = {0: 0, 1: 0}
+        server = {0: 0, 1: 0}
+        try:
+            for p in self._client_trace().get_1rtt(Direction.FROM_CLIENT):
+                client[int(p.key_phase)] += 1
+            for p in self._server_trace().get_1rtt(Direction.FROM_SERVER):
+                server[int(p.key_phase)] += 1
+        except Exception:
+            logging.info(
+                "Failed to read key phase bits. Potentially incorrect SSLKEYLOG?"
+            )
+            return TestResult.FAILED
+
+        succeeded = client[0] * client[1] * server[0] * server[1] > 0
```

Comment:
cute

Reply:
But what about integer overflow?

Reply:
Python does some weird stuff in the background to pretend that overflows don't exist.
Try running the following code:
```python
for i in range(1, 500):
    print(2**i)
```

Reply:
Oh, that's right.  Python numbers are all bignums.